pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -159,13 +159,32 @@ func NewServiceDiscovery(
 	return s
 }
 
+// merge workloadentry metadata.labels with spec.labels, with precedence to the metadata.
+func populateWorkloadEntryLabels(cfg config.Config) {
+	wle := cfg.Spec.(*networking.WorkloadEntry)
+	if wle == nil {
+		return
+	}
+	labels := make(map[string]string, len(wle.Labels)+len(wle.Labels))
+	for k, v := range wle.Labels {
+		labels[k] = v
+	}
+	// we will merge labels from metadata with spec, with precedence to the metadata
+	for k, v := range cfg.Labels {
+		labels[k] = v
+	}
+	wle.Labels = labels
```

Comment:
There's data race [as the test shows](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/37894/unit-tests_istio/1503205381462036480), maybe we can merge labels in `getWorkloadServiceEntries` and `convertWorkloadEntryToServiceInstances` (using config instead of WLE, like `convertWorkloadEntryToWorkloadInstance`) rather than modifying the WLE directly

Directionally I think it makes sense because it unifies WLE label handling

Reply:
I want to merge labels once originally.

Reply:
Or you may create a shallow copied WLE whose labels are merged, either is fine
```go
func populateWorkloadEntryLabels(cfg config.Config) *networking.WorkloadEntry {
	wle := cfg.Spec.(*networking.WorkloadEntry)
	if wle == nil {
		return nil
	}
	labels := make(map[string]string, len(wle.Labels)+len(wle.Labels))
	for k, v := range wle.Labels {
		labels[k] = v
	}
	// we will merge labels from metadata with spec, with precedence to the metadata
	for k, v := range cfg.Labels {
		labels[k] = v
	}
	copy := *wle
	copy.Labels = labels
	return &copy
}
```

Reply:
good suggestion, this will prevent updating anywhere